src/Twilio/Security/RequestValidator.cs

```diff
@@ -124,23 +124,45 @@ private static bool SecureCompare(string a, string b)
             return mismatch == 0;
         }
 
-        private string RemovePort(UriBuilder uri)
+        private string RemovePort(string url)
         {
-            // UriBuilder.ToString() will not display the port 
-            // if the Port property is set to -1
-            uri.Port = -1;
-            return uri.ToString();
+            return SetPort(url, -1);
         }
 
-        private string AddPort(UriBuilder uri)
+        private string AddPort(string url)
         {
-            if (uri.Port != -1)
+            var uri = new UriBuilder(url);
+            return SetPort(url, uri.Port);
+        }
+
+        private string SetPort(string url, int port)
+        {
+            var uri = new UriBuilder(url);
+            uri.Host = PreserveHostnameCasing(url);
+            if (port == -1)
             {
-                return uri.ToString();
+                uri.Port = port;
+            } else {
+                uri.Port = uri.Scheme == "https" ? 443 : 80;
             }
-            uri.Port = uri.Scheme == "https" ? 443 : 80;
-            return uri.ToString();
+            var scheme = PreserveSchemeCasing(url);
+            return uri.Uri.OriginalString.Replace(uri.Scheme, scheme);
+        }
+
+        private string PreserveHostnameCasing(string url)
+        {
+            var uri = new UriBuilder(url);
+            var startIndex = url.IndexOf(uri.Host, StringComparison.OrdinalIgnoreCase);
+            return url.Substring(startIndex, uri.Host.Length);
         }
 
+        // uri.Scheme will lowercase the output
+        private string PreserveSchemeCasing(string url)
+        {
+            var uri = new UriBuilder(url);
+            var startIndex = url.IndexOf(uri.Scheme, StringComparison.OrdinalIgnoreCase);
+            var preservedScheme = url.Substring(startIndex, uri.Scheme.Length);
+            return preservedScheme;
+        }
```

Comment:
This can be refactored to a single function which takes in the url and the part needing to be replaced: `PreserveCasing(url, uri.Host)` and `PreserveCasing(url, uri.Scheme)`.